arwen/elrondapi/elrondei.go

```diff
@@ -1402,23 +1400,20 @@ func v1_3_upgradeContract(
 
 	// Set up the async call as if it is not known whether the called SC
 	// is in the same shard with the caller or not. This will be later resolved
-	// in the handler for BreakpointAsyncCall.
-	codeEncoded := hex.EncodeToString(code)
-	codeMetadataEncoded := hex.EncodeToString(codeMetadata)
-	finalData := arwen.UpgradeFunctionName + "@" + codeEncoded + "@" + codeMetadataEncoded
+	// by runtime.ExecuteAsyncCall().
+	callData := txDataBuilder.NewBuilder()
+	callData.Func(arwen.UpgradeFunctionName)
+	callData.Bytes(code).Bytes(codeMetadata)
+
 	for _, arg := range data {
-		finalData += "@" + string(arg)
+		callData.Bytes(arg)
 	}
 
-	runtime.SetAsyncCallInfo(&arwen.AsyncCallInfo{
-		Destination: calledSCAddress,
-		Data:        []byte(finalData),
-		GasLimit:    uint64(gasLimit),
-		ValueBytes:  value,
-	})
-
-	// Instruct Wasmer to interrupt the execution of the caller SC.
-	runtime.SetRuntimeBreakpointValue(arwen.BreakpointAsyncCall)
+	runtime.ExecuteAsyncCall(
+		calledSCAddress,
```

Comment:
Reading it again and I think this should have been handled with ExecuteOnDestContext. Not with asyncCall

Reply:
the contract could in principle be in another shard, if "change owner" was called

Reply:
`ExecuteOnDestContext` is safe as long as we don't migrate accounts between shards.

Reply:
I mean, the contract could have its ownership transferred to an account from another shard. The new owner will be the one making the update